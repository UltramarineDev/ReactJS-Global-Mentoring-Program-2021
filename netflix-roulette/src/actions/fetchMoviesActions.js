import { tabs } from 'components/shared/MoviesList/constants';

import * as constants from '../constants';

const getMovieSuccessAction = (movie) => ({ type: constants.GET_MOVIE_SUCCESS, payload: movie });

const getMoviesSuccessAction = (movies) => ({ type: constants.GET_MOVIES_SUCCESS, payload: movies });

const getMoviesErrorAction = (error) => ({ type: constants.GET_MOVIES_ERROR, payload: error });

const getMoviesPendingAction = () => ({ type: constants.GET_MOVIES_PENDING });

const getMoviePendingAction = () => ({ type: constants.GET_MOVIE_PENDING });

export const getMoviesAction = (
  filter = undefined,
  sortBy = constants.DEFAULT_SORT_BY,
  search = undefined,
) => async (dispatch) => {
  try {
    dispatch(getMoviesPendingAction());
    let url = `${constants.URL}?sortBy=${sortBy}&sortOrder=${constants.DEFAULT_SORT_ORDER}&limit=${constants.DEFAULT_LIMIT}`;
    if (filter && filter !== tabs.ALL) {
      url = `${url}&filter=${filter}`;
    }

    if (search) {
      url = `${url}&search=${search}&searchBy=title`;
    }

    await fetch(url).then((res) => res.json()).then((res) => dispatch(getMoviesSuccessAction(res.data)));
  } catch (error) {
    dispatch(getMoviesErrorAction(error));
  }
};

export const getMovieAction = (id) => async (dispatch) => {
  try {
    dispatch(getMoviePendingAction());
    const url = `${constants.URL}/${id}`;
    await fetch(url).then((res) => res.json()).then((res) => dispatch(getMovieSuccessAction(res)));
  } catch (error) {
    dispatch(getMoviesErrorAction(error));
  }
};
