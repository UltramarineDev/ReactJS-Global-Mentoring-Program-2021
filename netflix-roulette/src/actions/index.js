export { getMoviesAction, getMovieAction } from './fetchMoviesActions';
export { addMovieAction, updateMovieAction, deleteMovieAction } from './movieActions';
export { setSearchExpressionAction } from './searchActions';
