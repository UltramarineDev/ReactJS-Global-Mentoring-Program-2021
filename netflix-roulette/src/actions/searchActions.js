import * as constants from '../constants';

export const setSearchExpressionAction = (search) => ({
  type: constants.SET_SEARCH_EXPRESSION,
  payload: search,
});
