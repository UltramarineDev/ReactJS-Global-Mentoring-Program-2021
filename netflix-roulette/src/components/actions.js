import * as constants from './constants';

const getMoviesSuccessAction = movies => 
  ({ type: constants.GET_MOVIES, payload: movies });

const getMovieSuccessAction = movie => 
  ({ type: constants.GET_MOVIE, payload: movie });

export const getMoviesAction = (filter = undefined, sortBy = constants.DEFAULT_SORT_BY) => async (dispatch) => {
  let url = `${constants.URL}?sortBy=${sortBy}&sortOrder=${constants.DEFAULT_SORT_ORDER}&limit=${constants.DEFAULT_LIMIT}`;

  if (filter) {
    url = `${url}&filter=${filter}`;
  }
  const response = await fetch(url).then(res => res.json());
  dispatch(getMoviesSuccessAction(response.data));
};

export const getMovieAction = (id) => async (dispatch) => {
   const response = await fetch(`${constants.URL}/${id}`).then(res => res.json());
   dispatch(getMovieSuccessAction(response));
};

export const createMovieAction = (input) => async (dispatch) => {
  const response = await fetch(`${constants.URL}`, {
      body: input,
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
    });
};

export const updateMovieAction = (input) => async (dispatch) => {
  const response = await fetch(`${constants.URL}`, {
      body: input,
      method: 'PUT',
      headers: {
        'content-type': 'application/json',
      },
    });
  };
