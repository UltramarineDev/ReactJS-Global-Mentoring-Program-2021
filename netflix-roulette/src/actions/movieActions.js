import * as constants from '../constants';
import { getMoviesAction } from './fetchMoviesActions';

const deleteMovieCompletedAction = () => ({ type: constants.DELETE_MOVIE });

const deleteMoviePendingAction = () => ({ type: constants.DELETE_MOVIE_PENDING });

export const addMovieAction = (input) => async (dispatch) => {
  await fetch(`${constants.URL}`, {
    body: JSON.stringify(input),
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  }).then(dispatch(getMoviesAction()));
};

export const updateMovieAction = (input) => async (dispatch) => {
  await fetch(`${constants.URL}`, {
    body: JSON.stringify(input),
    method: 'PUT',
    headers: {
      Accept: 'application/json',
      'content-type': 'application/json',
    },
  }).then(dispatch(getMoviesAction()));
};

export const deleteMovieAction = (movieId) => async (dispatch) => {
  if (!movieId) {
    return;
  }

  try {
    dispatch(deleteMoviePendingAction());
    await fetch(`${constants.URL}/${movieId}`, {
      method: 'DELETE',
      headers: {
        Accept: 'application/json',
        'content-type': 'application/json',
      },
    }).then(dispatch(deleteMovieCompletedAction()));
  } catch (error) {
    console.log('delete movie error');
  }
};
