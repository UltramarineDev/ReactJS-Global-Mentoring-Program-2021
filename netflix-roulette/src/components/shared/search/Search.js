import React, { useState, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { useRouter } from 'next/router'
import { buttonTypes, buttonSizes } from '../constants';
import { wordings } from '../../../locales/wordings';
import { setSearchExpressionAction } from '../../../actions';

import Button from '../Button/Button';
import SearchField from './SearchField/SearchField';
import styles from '../../../../styles/Search.module.scss';

const Search = () => {
  const router = useRouter();
  const { query } = router.query;
  const [searchExpression, setSearchExpression] = useState(query);
  const dispatch = useDispatch();

  useEffect(() => {
    if (searchExpression) {
      dispatch(setSearchExpressionAction(searchExpression));
    }
  }, [searchExpression]);

  const handleSearch = () => {
    dispatch(setSearchExpressionAction(searchExpression));
    searchExpression ? router.push(`/search/${searchExpression}`) : router.push('/');
  };

  const handleChange = (e) => {
    const { value } = e.target;
    setSearchExpression(value);
  };

  return (
    <div className={styles.search}>
      <SearchField onChange={handleChange} initialValue={searchExpression} />
      <Button label={wordings.search} type={buttonTypes.PRIMARY} size={buttonSizes.LARGE} onClick={handleSearch} />
    </div>
  );
};

export default Search;
