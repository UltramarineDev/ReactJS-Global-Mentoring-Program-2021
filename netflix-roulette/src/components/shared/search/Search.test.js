import React from 'react';
import { cleanup, render, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import { setSearchExpressionAction } from 'actions';

import Search from './Search';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useDispatch: () => mockDispatch,
}));

jest.mock('actions', () => ({
  setSearchExpressionAction: jest.fn() }
));

describe('Search tests', () => {
  afterEach(() => {
    cleanup();
  });
  test('renders Search', () => {
    const component = renderer.create(
      <MemoryRouter>
        <Search />
      </MemoryRouter>,
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('tests handleChange Search', () => {
    const { getByLabelText } = render(
      <MemoryRouter>
        <Search />
      </MemoryRouter>,
    );

    const searchField = getByLabelText('search_field');
    fireEvent.change(searchField, { target: { value: '23' } });

    expect(setSearchExpressionAction).toBeCalled();
  });
});
