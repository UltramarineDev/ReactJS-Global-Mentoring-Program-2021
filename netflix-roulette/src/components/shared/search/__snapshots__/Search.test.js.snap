// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search tests renders Search 1`] = `
<div
  className="search"
>
  <input
    aria-label="search_field"
    onChange={[Function]}
    placeholder="What do you want to watch?"
    type="search"
    value=""
  />
  <button
    className="searchButton largeSize"
    data-testid="save_form"
    onClick={[Function]}
    type="submit"
  >
    <span
      className="labelSearch"
    >
      SEARCH
    </span>
  </button>
</div>
`;
